### PR TITLE
Read notebook from stdin/write to stdout when path is '-'

### PR DESCRIPTION
**Pull request by mwouts**
This pull requests implements `-` as stdin/stdout, cf. #372.

A case when one wants to read notebooks from stdin are Jupytext notebooks (#365). Piping notebooks to stdout can also be useful in case of chained notebook conversions (to html with `nbconvert` for instance).

I have included two tests. What remains is probably to
- [x] review the stdout option, and the tests
- [ ] mention `-` as stdin/stdout in Papermill's documentation, or at least in the CLI
- [x] increase the default value for `--log-level` when the output path is `-`

I have tested the PR with the command below, which prints the notebook (represented as an Hydrogen script) once the injected parameter cell as been added by papermill:
```
jupytext notebook_as_script.py --to ipynb -o - | papermill - - --prepare-only -p a 2 --log-level ERROR | jupytext --to py:percent
```
where `notebook_as_script.py` is a notebook represented as a script using Jupytext, e.g.:
```
# ---
# jupyter:
#   jupytext:
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.2'
#       jupytext_version: 1.1.3
#   kernelspec:
#     display_name: Python3
#     language: python
#     name: python3
# ---

# %% {"tags": ["parameters"]}
a = 1

# %%
print('a={}'.format(a))
```

As expected, the command above returns the updated representation of the notebook, with an additional cell in between:
```
# %% {"tags": ["injected-parameters"]}
# Parameters
a = 2
```

### CODE REVIEW

**Comment by MSeal** [comment]
Thanks for making the PR! I agree with your list of TODO items. One thing to note is that the logger is defaulting to stderr so it shouldn't impact piping results, but I agree that when piping out it's probably a better default expectation.

I'd add one quality of life feature as well. If we're reading/writing from a pipe we should default the input / output paths to `"-"`. Thus `jupytext notebook_as_script.py --to ipynb -o - | papermill --prepare-only -p a 2 | jupytext --to py:percent` could work too. Or just `jupytext notebook_as_script.py --to ipynb -o - | papermill output.ipynb --prepare-only -p a 2`.

I believe adding this to cli.py would achieve this effect:
```
import os
import sys
from stat import S_ISFIFO
...

  # Check if input is being piped
  if num_path_args < 2 and S_ISFIFO(os.fstat(0).st_mode):
    notebook_path = '-'
    num_path_args += 1
  # Check if output it being piped
  if num_path_args < 2 and not sys.stdout.isatty():
    output_path = '-'
```
and 
```
  if output_path == '-': 
    ... # Change default log level?
```
afterwards to capture log defaulting after path defaulting.

**Comment by mwouts** [comment]
Thanks @MSeal, sure I'll do as you suggest -  thanks by the way for pointing out to this possibility, I'm glad to learn about this... in the end piping will become very concise!

**Comment by mwouts** [comment]
@MSeal , I have pushed an additional commit, with which I have been able to pipe in/out notebooks with no unnecessary argument: 
```
jupytext notebook_as_script.py --to ipynb | papermill --prepare-only -p a 3 | jupytext --to py
```

I would have liked to know more about `click`, but unfortunately that is not the case... So I've done this directly in the `papermill` function. 

If you provide more hints on how to do it with `click`, I'd be happy to change the implementation. I've seen the `callback` argument in `click.argument`, but what looked non trivial was a) to exchange the two arguments if only one is provided and b) to raise an error when no value is given, and we're not piping (which is an issue also in the current proposal - would you like to raise an exception in that case?).

**Comment by rgbkrk** [comment]
Oooh, I like that approach to check if it's not a tty and allow no args.

File: papermill/cli.py [hunk 1]
```diff
@@ -29,8 +33,8 @@ def print_papermill_version(ctx, param, value):
 
 
 @click.command(context_settings=dict(help_option_names=['-h', '--help']))
-@click.argument('notebook_path')
-@click.argument('output_path')
+@click.argument('notebook_path', required=False)
+@click.argument('output_path', required=False)
```
Comment (rgbkrk) [inline papermill/cli.py]: The only awkward thing here is that it's required if not piped, yet there's likely not a way to specify this for click. It's like we want to require it yet override it during one of the cases below (somehow before?).

Comment (MSeal) [reply papermill/cli.py]: You can do the `INPUT_PIPED = S_ISFIFO(os.fstat(0).st_mode)` and `OUTPUT_PIPED = sys.stdout.isatty()` outside the function set `required` based on those variables is probably the best tradeoff.

File: papermill/cli.py [hunk 2]
```diff
@@ -129,6 +133,22 @@ def papermill(
     output in the destination notebook.
 
     """
+    # Check if input is being piped
+    if (not notebook_path) or (not output_path) and S_ISFIFO(os.fstat(0).st_mode):
```
Comment (MSeal) [inline papermill/cli.py]:
```suggestion
if (not notebook_path or not output_path) and S_ISFIFO(os.fstat(0).st_mode)
```
of the logic won't work quite correctly.

Comment (MSeal) [reply papermill/cli.py]: Probably means we need a unittest for this check too

**Comment by MSeal** [comment]
@mwouts We could get really tricky with click, but I think with the one suggested change to `required=` the code will be easy to read and understand. I would just keep the `if piped: ... path = '-'` pattern you have as other maintainers won't need to learn tricky library calls to manipulate the logic.

File: papermill/cli.py (hunk 1 above)
Comment (mwouts) [reply papermill/cli.py]: I agree. Do you want something like the below? It seems to work in every case, but for a detail: when an input pipe exists, the two arguments become optional.
```
import os
import sys
from stat import S_ISFIFO
import click

INPUT_PIPED = S_ISFIFO(os.fstat(0).st_mode)
OUTPUT_PIPED = sys.stdout.isatty()


@click.command(context_settings=dict(help_option_names=['-h', '--help']))
@click.argument('notebook_path', required=not INPUT_PIPED)
@click.argument('output_path', required=not (INPUT_PIPED or OUTPUT_PIPED))
def papermill(notebook_path, output_path):
    if INPUT_PIPED and notebook_path and not output_path:
        output_path = notebook_path
        notebook_path = '-'
    else:
        notebook_path = notebook_path or '-'
        output_path = output_path or '-'

    print('notebook_path={}'.format(notebook_path))
    print('output_path={}'.format(output_path))


if __name__ == '__main__':
    papermill()
```

Comment (MSeal) [reply papermill/cli.py]: Yeah that about what I was expecting. And I think the logic for required there is correct to match expected behavior.

On a related note we'll want to test this on a variety of OS's despite those methods not having explicit warnings. I can check the windows / linux setups easily enough.

File: papermill/cli.py (hunk 2 above)
Comment (mwouts) [reply papermill/cli.py]: Interesting! Apparently I misunderstood what my editor told me... (anyway, as discussed this part goes away with the next commit)

image

**Comment by mwouts** [comment]
> On a related note we'll want to test this on a variety of OS's despite those low-level os methods not having explicit warnings. I can check the windows / ubuntu setups easily enough, so just need to check osx for the big 3.

I have been working on automated tests at https://github.com/mwouts/papermill/commit/a506a26659a0e6f2b9f20ea2af1c5d1eb817e51f.

Currently 2 out of 3 tests are not passing on Windows (Win 10, Python 3.6 from Anaconda). Interestingly, when I pipe on the command line I get
- one progress bar
- the original notebook
- the output notebook written twice.

I can fix the progress bar by setting `progress_bar=False` when output is a pipe. However, I am a bit surprise to see a total of three notebooks. Is the input notebook written back by papermill? Is the output notebook really written twice?

To reproduce: create `nb.ipynb` as 
```
{
 "cells": [],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
```
and execute `type nb.ipynb | papermill`
```
0it [00:00, ?it/s]{
 "cells": [],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "papermill": {
   "duration": null,
   "end_time": null,
   "environment_variables": {},
   "exception": null,
   "input_path": "-",
   "output_path": "-",
   "parameters": {},
   "start_time": "2019-06-04T05:34:02.966635",
   "version": "1.0.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}{
 "cells": [],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "papermill": {
   "duration": 2.308568,
   "end_time": "2019-06-04T05:34:05.275203",
   "environment_variables": {},
   "exception": null,
   "input_path": "-",
   "output_path": "-",
   "parameters": {},
   "start_time": "2019-06-04T05:34:02.966635",
   "version": "1.0.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
{
 "cells": [],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "papermill": {
   "duration": 2.308568,
   "end_time": "2019-06-04T05:34:05.275203",
   "environment_variables": {},
   "exception": null,
   "input_path": "-",
   "output_path": "-",
   "parameters": {},
   "start_time": "2019-06-04T05:34:02.966635",
   "version": "1.0.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
```

**Comment by mwouts** [comment]
I see that the notebook is indeed saved many times - possibly each time a cell gets executed. We could change that by setting `output_path=None if output_path == '-' else output_path` at [execute.py:92](https://github.com/nteract/papermill/blob/2540bc663f8524f8d31764e4574e45213f17b827/papermill/execute.py#L92).

@MSeal , what do you think? Maybe we should even go a bit further and introduce an additional argument, say `save_progress` or `save_at_every_cell` (`True` by default) in `execute_notebook` and `papermill` CLI?

**Comment by MSeal** [comment]
Oh yes. I completely forgot about this. Yes if the output notebook path is a tty we do need an extra option to control save rates. `request_save_on_cell_execute` would be most appropriate, as not all engines can in theory actually save after each cell executes so it should be an option that states that will always happen.

**Comment by mwouts** [comment]
Good - I'll propose this soon. By the way, could you also let me know what you think of the [tentative tests](https://github.com/mwouts/papermill/commit/a506a26659a0e6f2b9f20ea2af1c5d1eb817e51f)? Would you like to see something like that in this PR?

**Comment by MSeal** [comment]
Yes I think tests like that would be good, you should also include the `-` versions of each input / output path. Also the subprocess should be careful to use the cli being tested and not a different installation of `papermill`. `Popen(sys.executable + ' -m papermill.cli + <rest of args here>')` should do the trick well enough for test environment contexts. Pytest will handle preserving the `PYTHONPATH` which the subprocess should inherit with Popen.

**Comment by mwouts** [comment]
@MSeal , I think we're getting closer! The tests now pass both on Windows and Linux. Do you see anything that should be added? Can I let you add a word about this in the documentation?

**Comment by MSeal** [comment]
@rgbkrk Could you give the PR code a test run on osx with and without the tty patterns to make sure there's no os specific oddities you hit with these changes?

**Comment by rgbkrk** [comment]
Yeah I can do that!

**Comment by rgbkrk** [comment]
The tests for `test_pipe_in_out_auto` and `test_pipe_in_out_explicit` fail for me on Python 2.7 and pass in Python 3.

<details>
<summary>Expand for output for failed tests</summary>

```
papermill/tests/test_cli.py::test_pipe_in_out_auto FAILED                                                                                                                     [ 19%]
papermill/tests/test_cli.py::test_pipe_in_out_explicit FAILED                                                                                                                 [ 20%]

===================================================================================== FAILURES ======================================================================================
_______________________________________________________________________________ test_pipe_in_out_auto _______________________________________________________________________________

notebook = {'nbformat_minor': 2, 'cells': [{'source': u'This is a notebook with kernel: J...e': u'Javascript (Node.js)', 'name': 'javascript', 'language': u'javascript'}}}

    @require_papermill_installed
    def test_pipe_in_out_auto(notebook):
        process = papermill_cli(stdout=subprocess.PIPE, stdin=subprocess.PIPE)
        text = nbformat.writes(notebook)
        out, err = process.communicate(input=text.encode('utf-8'))

        # Test no message on std error
        assert not err

        # Test that output is a valid notebook
>       nbformat.reads(out.decode('utf-8'), as_version=4)

papermill/tests/test_cli.py:698:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py27/lib/python2.7/site-packages/nbformat/__init__.py:74: in reads
    nb = reader.reads(s, **kwargs)
.tox/py27/lib/python2.7/site-packages/nbformat/reader.py:58: in reads
    nb_dict = parse_json(s, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

s = '', kwargs = {}

    def parse_json(s, **kwargs):
        """Parse a JSON string into a dict."""
        try:
            nb_dict = json.loads(s, **kwargs)
        except ValueError:
            # Limit the error message to 80 characters.  Display whatever JSON will fit.
>           raise NotJSONError(("Notebook does not appear to be JSON: %r" % s)[:77] + "...")
E           NotJSONError: Notebook does not appear to be JSON: u''...

.tox/py27/lib/python2.7/site-packages/nbformat/reader.py:17: NotJSONError
------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------
Failed to run command:
[u'ijskernel', u'--hide-undefined', u'/var/folders/kd/cylz4mhs1_9cpsjh0_c_gzfr0000gn/T/tmpAHOPLM.json', u'--protocol=5.0']
    PATH='/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/bin:/Users/kylek/.deno/bin:/Users/kylek/miniconda3/bin:/Users/kylek/.nvm/versions/node/v10.15.3/bin:./bin:/usr/local/bin:/usr/local/sbin:/Users/kylek/.dotfiles/bin:/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/bin:/Users/kylek/.rbenv/bin:/Users/kylek/.rbenv/shims:/Users/kylek/code/bin'
    with kwargs:
{'cwd': None, 'stdin': -1, 'preexec_fn': <function <lambda> at 0x10d1c92a8>, 'stderr': None, 'stdout': None}

Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/kylek/code/src/github.com/nteract/papermill/papermill/cli.py", line 231, in <module>
    papermill()
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/kylek/code/src/github.com/nteract/papermill/papermill/cli.py", line 191, in papermill
    cwd=cwd,
  File "papermill/execute.py", line 100, in execute_notebook
    **engine_kwargs
  File "papermill/engines.py", line 58, in execute_notebook_with_engine
    return self.get_engine(engine_name).execute_notebook(nb, kernel_name, **kwargs)
  File "papermill/engines.py", line 310, in execute_notebook
    nb = cls.execute_managed_notebook(nb_man, kernel_name, log_output=log_output, **kwargs)
  File "papermill/engines.py", line 372, in execute_managed_notebook
    preprocessor.preprocess(nb_man, safe_kwargs)
  File "papermill/preprocess.py", line 18, in preprocess
    with self.setup_preprocessor(nb_man.nb, resources, km=km):
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/nbconvert/preprocessors/execute.py", line 324, in setup_preprocessor
    self.km, self.kc = self.start_new_kernel(cwd=path)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/nbconvert/preprocessors/execute.py", line 271, in start_new_kernel
    km.start_kernel(extra_arguments=self.extra_arguments, **kwargs)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/jupyter_client/manager.py", line 259, in start_kernel
    **kw)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/jupyter_client/manager.py", line 204, in _launch_kernel
    return launch_kernel(kernel_cmd, **kw)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/jupyter_client/launcher.py", line 138, in launch_kernel
    proc = Popen(cmd, **kwargs)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
_____________________________________________________________________________ test_pipe_in_out_explicit _____________________________________________________________________________

notebook = {'nbformat_minor': 2, 'cells': [{'source': u'This is a notebook with kernel: J...e': u'Javascript (Node.js)', 'name': 'javascript', 'language': u'javascript'}}}

    @require_papermill_installed
    def test_pipe_in_out_explicit(notebook):
        process = papermill_cli(['-', '-'], stdout=subprocess.PIPE, stdin=subprocess.PIPE)
        text = nbformat.writes(notebook)
        out, err = process.communicate(input=text.encode('utf-8'))

        # Test no message on std error
        assert not err

        # Test that output is a valid notebook
>       nbformat.reads(out.decode('utf-8'), as_version=4)

papermill/tests/test_cli.py:711:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py27/lib/python2.7/site-packages/nbformat/__init__.py:74: in reads
    nb = reader.reads(s, **kwargs)
.tox/py27/lib/python2.7/site-packages/nbformat/reader.py:58: in reads
    nb_dict = parse_json(s, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

s = '', kwargs = {}

    def parse_json(s, **kwargs):
        """Parse a JSON string into a dict."""
        try:
            nb_dict = json.loads(s, **kwargs)
        except ValueError:
            # Limit the error message to 80 characters.  Display whatever JSON will fit.
>           raise NotJSONError(("Notebook does not appear to be JSON: %r" % s)[:77] + "...")
E           NotJSONError: Notebook does not appear to be JSON: u''...

.tox/py27/lib/python2.7/site-packages/nbformat/reader.py:17: NotJSONError
------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------
Failed to run command:
[u'ijskernel', u'--hide-undefined', u'/var/folders/kd/cylz4mhs1_9cpsjh0_c_gzfr0000gn/T/tmpSiHSKe.json', u'--protocol=5.0']
    PATH='/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/bin:/Users/kylek/.deno/bin:/Users/kylek/miniconda3/bin:/Users/kylek/.nvm/versions/node/v10.15.3/bin:./bin:/usr/local/bin:/usr/local/sbin:/Users/kylek/.dotfiles/bin:/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/bin:/Users/kylek/.rbenv/bin:/Users/kylek/.rbenv/shims:/Users/kylek/code/bin'
    with kwargs:
{'cwd': None, 'stdin': -1, 'preexec_fn': <function <lambda> at 0x108c9c2a8>, 'stderr': None, 'stdout': None}

Traceback (most recent call last):
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/kylek/code/src/github.com/nteract/papermill/papermill/cli.py", line 231, in <module>
    papermill()
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/kylek/code/src/github.com/nteract/papermill/papermill/cli.py", line 191, in papermill
    cwd=cwd,
  File "papermill/execute.py", line 100, in execute_notebook
    **engine_kwargs
  File "papermill/engines.py", line 58, in execute_notebook_with_engine
    return self.get_engine(engine_name).execute_notebook(nb, kernel_name, **kwargs)
  File "papermill/engines.py", line 310, in execute_notebook
    nb = cls.execute_managed_notebook(nb_man, kernel_name, log_output=log_output, **kwargs)
  File "papermill/engines.py", line 372, in execute_managed_notebook
    preprocessor.preprocess(nb_man, safe_kwargs)
  File "papermill/preprocess.py", line 18, in preprocess
    with self.setup_preprocessor(nb_man.nb, resources, km=km):
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/nbconvert/preprocessors/execute.py", line 324, in setup_preprocessor
    self.km, self.kc = self.start_new_kernel(cwd=path)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/nbconvert/preprocessors/execute.py", line 271, in start_new_kernel
    km.start_kernel(extra_arguments=self.extra_arguments, **kwargs)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/jupyter_client/manager.py", line 259, in start_kernel
    **kw)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/jupyter_client/manager.py", line 204, in _launch_kernel
    return launch_kernel(kernel_cmd, **kw)
  File "/Users/kylek/code/src/github.com/nteract/papermill/.tox/py27/lib/python2.7/site-packages/jupyter_client/launcher.py", line 138, in launch_kernel
    proc = Popen(cmd, **kwargs)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 394, in __init__
    errread, errwrite)
  File "/usr/local/Cellar/python@2/2.7.16/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1047, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

</details>

The kernel errors seem real suspect though, since I do have a bad js kernel on disk right now.

**Comment by rgbkrk** [comment]
This worked like a champ:

```
cat cli_example.ipynb | papermill > test.ipynb
```

As did these

```
$ cat cli_example.ipynb | papermill - what.ipynb
$ cat cli_example.ipynb | papermill - -
```

I think one weird thing about this setup though is that `output_path` is encoded into the papermill metadata. When it's tty mode, this appears as `-`. It's informative to us as the developers, I would hope that doesn't confuse end users.

**Comment by rgbkrk** [comment]
I'm going to clean up my kernels now that I see the issue is entirely related to my bad kernel on disk, and python 3 just happened to shuffle the dict the right way on the first try. 😉

**Comment by rgbkrk** [comment]
This works beautifully on my mac. I'm strongly in favor of the nice unix composability here. ❤️ 🤤

**Comment by MSeal** [comment]
Thanks for the dev work here folks!

**Comment by mwouts** [comment]
That's great! Thanks @MSeal and @rgbkrk for your support, it was very interesting to work on this.

**Comment by rgbkrk** [comment]
I love that I can now chain things and pipe the output to `jq`:

```
$ cat original.ipynb | papermill -p a 3 | papermill -p a 5 | jq '.cells[].outputs'
[]
[]
[
  {
    "name": "stdout",
    "output_type": "stream",
    "text": [
      "5\n"
    ]
  }
]
[]
```

**Comment by mwouts** [comment]
@rgbkrk , you may also like piping into jupytext - that's a concise way to see the inputs and cell metadata (not the outputs, though!)
```
$ cat original.ipynb | papermill --prepare-only -p a 3 | jupytext --to md
```

**Pull request merged**